### ISSUE TITLE
Node: streams should emit "error" or "end", but not both

### ISSUE THREAD

**Issue by stephenplusplus** [grpc/grpc#8954]
*Originally posted: https://github.com/GoogleCloudPlatform/google-cloud-node/issues/1846#issuecomment-264466357*

It looks like a gRPC stream will emit both an `error` and an `end` event. While the stream is technically "ended" after an error occurs-- maybe more accurately, "closed"-- the `end` event should not be emitted.

```js
require('fs')
  .createReadStream('non-existent-file')
  .on('error', function(err) { console.log('emitted') })
  .on('end', function() { console.log('not emitted') })
  .on('data', function() {}) // to drain the data

require('request')
  .get('http://www.non-exitent-url.com')
  .on('error', function(err) { console.log('emitted') })
  .on('end', function() { console.log('not emitted') })
  .on('data', function() {}) // to drain the data
```

It's expected that the stream will either emit `error` or `end`, but not both. The gRPC stream will emit `end` first, then `error`.

I haven't looked deeply into the implementation in gRPC, but making a generalization; if this is the process:

- allow the stream the user is holding to complete its lifecycle
- do some post-processing
- emit `status`
- if there was an error, emit `error`

Consider using an intermediary stream, such as a [`through`](https://github.com/rvagg/through2) stream or a Transform stream, then registering a `prefinish` event handler to `cork()` the stream while you do the post processing. If after determining there was no error, you can call `uncork()` on the through stream, which will allow the end event to fire.

**Comment by murgatroid99** [comment]
This seems like a reasonable request. But it also looks like an API breakage, and I really want to avoid API breakages if at all possible. And there's no indication in the Stream API documentation that `error` and `end` are supposed to be mutually exclusive, so I don't think I can reasonably call the current behavior a bug.

If you know of some documentation that indicates that custom streams should behave as you describe, that would help.

**Comment by stephenplusplus** [comment]
I posed this question in [a tweet](https://twitter.com/stephenplusplus/status/806565478806290432).

> readable emit 'end', writable emit 'finish'. in case of errors no official guarantees (some emit close, others, error)
> [end-of-stream](https://github.com/mafintosh/end-of-stream) handles most cases for you (non-trivial)
> \- @mafintosh

I understand the concern about breaking the API. And based on that feedback, it doesn't appear to be a bug. It is inconvenient for users, however. While they will have separate success and error handlers, they will still need to check to see if the other handler executed.

With things the way they are currently, it is very difficult to implement a retry strategy. For reference, [this is the code](https://github.com/GoogleCloudPlatform/google-cloud-node/pull/1872/files) we put together to handle the issue in GCN.

@jmuk -- was this a problem for you when implementing a retry strategy in GAX? Have you run into this before and maybe have a clever way to handle it?

**Comment by murgatroid99** [comment]
Actually, I think that can work. If we accept that API guarantees are explicitly or implicitly discarded when there is an error, then we can say that behavior changes when there is an error event don't actually break any API guarantees.

**Comment by stephenplusplus** [comment]
That works for me 👍

**Comment by stephenplusplus** [comment]
Thank you for accepting this change! It looks like the work has been merged in #9101-- is a new release including this expected by any certain time?

**Comment by murgatroid99** [comment]
grpc@1.1.0-pre1 has been published and it includes that change, so this should now be fixed.
